pkg/storage/bucket/s3/config.go

```diff
@@ -86,6 +87,7 @@ func (cfg *Config) RegisterFlagsWithPrefix(prefix string, f *flag.FlagSet) {
 	f.BoolVar(&cfg.Insecure, prefix+"s3.insecure", false, "If enabled, use http:// for the S3 endpoint instead of https://. This could be useful in local dev/test environments while using an S3-compatible backend storage, like Minio.")
 	f.StringVar(&cfg.SignatureVersion, prefix+"s3.signature-version", SignatureVersionV4, fmt.Sprintf("The signature version to use for authenticating against S3. Supported values are: %s.", strings.Join(supportedSignatureVersions, ", ")))
 	f.StringVar(&cfg.BucketLookupType, prefix+"s3.bucket-lookup-type", BucketAutoLookup, fmt.Sprintf("The s3 bucket lookup style. Supported values are: %s.", strings.Join(supportedBucketLookupTypes, ", ")))
+	f.BoolVar(&cfg.AWSSDKAuth, prefix+"s3.aws-sdk-auth", false, "If enabled, the client will use default authentication methods of the AWS GO SDK for go based on known environment variables and known AWS config files.")
```

Comment:
default should be true?

If this is false we are changing the behavior as far as i understood?

Reply:
I think this is default is false in `minio`

Reply:
actually, after looking at minio's code a bit I am wondering if we should just "hardcode" `AWSSDKAuth: true` and not to have yet-another-configuration. Let me take a minio's code close to see if there is any cred resolution strategy that is not covered by `AWSSDKAuth` before its introduction.

Reply:
After reading https://github.com/thanos-io/thanos/pull/4667 and the code I think it's safe to just use `AWSSDKAuth: true` and don't expose yet-another-config.  I am not sure what would be the benefit of using the the cred providers from `minio` over aws go sdk v2; I think minio didn't use aws sdk to minimize dependencies? 

Also, we will need to have some logic to set `AWSSDKAuth: false` if `AccessKeyID` is defined; they can't co-exist.

Reply:
I agree these options are hard to decide by users and we should try to avoid them. I am happy you jump on this quick :)

But here is the thing: s3 provider in objstore supports `AWSSDKAuth: false` and `AccessKey: ""`.

https://github.com/thanos-io/objstore/blob/cec51c61948b0c04b60c035fd372cc83a540881b/providers/s3/s3.go#L231-L242

this is the code path my configuration was using in AWS and it was just by chance working due to their bug.

This makes no sense to AWS, but it makes sense to minio. Because minio has support for access without AccessKey that sort of works like AWS but doesn't use AWS libraries. See https://min.io/docs/minio/linux/developers/security-token-service.html

So I think the other possible solution to this is to rename our current [s3](https://github.com/cortexproject/cortex/tree/master/pkg/storage/bucket/s3) provider to minio provider and have a native s3 provider that sets AWSSDKAuth: true or even better don't use any minio at all. In that sense, Thanos should also do the same, I think.

I like this option better because it simplifies the use case for AWS and still enables minio.

Let me know what you think @alvinlin123 @alanprot @yeya24

Reply:
I agree these options are hard to decide by users and we should try to avoid them. I am happy you jump on this quick :)

But here is the thing: s3 provider in objstore supports `AWSSDKAuth: false` and `AccessKey: ""`.

https://github.com/thanos-io/objstore/blob/cec51c61948b0c04b60c035fd372cc83a540881b/providers/s3/s3.go#L231-L242

this is the code path my configuration was using in AWS and it was just by chance working due to their bug.

This makes no sense to AWS, but it makes sense to minio. Because minio has support for access without AccessKey that sort of works like AWS but doesn't use AWS libraries. See https://min.io/docs/minio/linux/developers/security-token-service.html

So I think the other possible solution to this is to rename our current [s3](https://github.com/cortexproject/cortex/tree/master/pkg/storage/bucket/s3) provider to minio provider and have a native s3 provider that sets AWSSDKAuth: true or even better don't use any minio at all. In that sense, Thanos should also do the same, I think.

I like this option better because it simplifies the use case for AWS and still enables minio.

Let me know what you think @alvinlin123 @alanprot @yeya24

Reply:
I think the resolver chain from Thanos at https://github.com/thanos-io/objstore/blob/cec51c61948b0c04b60c035fd372cc83a540881b/providers/s3/s3.go#L231-L242  is a subset of what the AWS SDK does. 

In other words, the following cred resolving order implemented in Minio is covered by the AWS SDK; AWS SDK supports assume role with web identity, env var, loading shared profile etc. 

```
chain = []credentials.Provider{
			wrapCredentialsProvider(&credentials.EnvAWS{}),
			wrapCredentialsProvider(&credentials.FileAWSCredentials{}),
			wrapCredentialsProvider(&credentials.IAM{
				Client: &http.Client{
					Transport: http.DefaultTransport,
				},
				Endpoint: config.STSEndpoint,
			}),
		}
```

So what works previously,  `AWSSDKAuth: false AccessKeyID:""`, should be equivalent to `AWSSDKAuth: true AccessKeyID:""` with the current version of Thanos and Minio.

Reply:
I see. AWS SDK might also do the right thing with minio in all the use cases, even without AccessKey.
Then we can skip this code path completely.

My use case is solved with AWSDKAuth: true, yes.  Let me implement the logic you mentioned before about nonempty AccessKey.